Core code generation

james-perretta opened the pull request:
- Emits bytecode for all language constructs (does not include definitions of builtin functions or the array bounds/type checking to go with them)
- Adds SCall, SQuote, SBool, and BoolLit types
- Removes Environment class in favor of dictionaries (since we only have local and global environments)

Fixes #3

james-perretta committed:
Wirking on adding call and quote to parser.

james-perretta committed:
Converted visitor tests from doctests to unit tests and added tests for new AST nodes.

james-perretta committed:
Updated environment assignment tests to account for SCall node.

james-perretta committed:
Fixed handling of quoted and unqoted symbols in IR generation.
Working on emitting function definitions.

james-perretta committed:
Implemented lambda code generation.

james-perretta committed:
Implemented global function call code gen.

james-perretta committed:
Wrote test for defining and calling a lambda in place.

james-perretta committed:
Wrote test for calling a function bound to a local variable.

james-perretta committed:
Implemented type_name on SBool and __str__ on BoolLit.

james-perretta committed:
Implemented emitting conditionals. Working on test cases.

james-perretta committed:
Finished writing tests for emitting conditionals.

james-perretta committed:
Removed Environment class and replaced uses with dictionaries.

james-perretta committed:
Removed empty test classes.

porglezomp commented:
Woo! Big review! I'll see how much of this I can get to tonight, but I'm also working on OS stuff.

porglezomp commented on bytecode.py:
```diff
@@ -67,20 +66,31 @@ def __str__(self) -> str:
 class EvalEnv:
     _local_env: Dict[Var, Value]
-    _global_env: Environment
+    _global_env: Dict[SSym, SExp]
```
Should this maybe be `Dict[SSym, Value]`?

james-perretta replied:
Sure

porglezomp commented on lint.sh:
```diff
@@ -4,5 +4,5 @@ status=0
 pycodestyle ./*.py || status=1
 # for some reason Travis-CI isort doesn't know about dataclasses???
-isort --diff --check-only --recursive -b dataclasses || status=1
+isort --diff --check-only --recursive -b dataclasses *.py || status=1
```
Doesn't `--recursive` find all relevant files automatically? I get warnings as I would expect without having to include this.

james-perretta replied:
Ah, so I have my virtual environment folder in the same directory as the py files, so I was getting a ton of errors in library files. I didn't mean to commit the change though and can undo it.

porglezomp commented on scheme.py:
```diff
@@ -132,6 +129,15 @@ def __str__(self) -> str:
+@dataclass(frozen=True)
+class Quote(SExp):
+    """A quoted list"""
+    slist: SList
```
Quotes are arbitrary s-expressions, not just lists. The most important non-list case is being able to write `'symbol`.

james-perretta replied:
Can you give an example of a program we want to be able to write that needs a standalone symbol literal?

james-perretta replied:
Nevermind, I'll update it

porglezomp commented on scheme.py:
```diff
@@ -208,7 +214,13 @@ def type_name(self) -> SSym:
-@dataclass
+@dataclass(frozen=True)
+class SCall(SExp):
+    func: Union[SSym, SFunction]
```
This doesn't allow for cases like the following:
```lisp
((if p fun1 fun2) args)
```
`func` should be an arbitrary s-expression.

james-perretta replied:
Updated and added a test case

james-perretta committed:
Updated parser to allow quoting symbols.
Updated SCall to allow arbitrary expressions for the procedure to be
called.
Updated global env mapping to only allow Value rather than any SExp.
Added --skip venv to isort flags, removed *.py.

james-perretta committed:
Added address() override for SBool.

porglezomp commented on emit_IR.py:
```diff
@@ -0,0 +1,243 @@
+        count += 1
+
+
+class EnvBindingNotFound(Exception):
```
When I was reading custom exceptions stuff, I found things suggesting that you want all of your exceptions in one file, which is why we have the `errors.py`. I *think* this should go there?

james-perretta replied:
I'm familiar with two philosophies: having all the exceptions in one place, and having exceptions defined in the same module that they logically "originate" from. If we want to go with all the exceptions in errors.py that's fine with me. 

We should probably also have a custom base class for our custom exceptions, no?

james-perretta committed:
Moved EnvBindingNotFound exception to errors.py.
